src/lib.rs

```diff
@@ -127,6 +129,42 @@ impl Multiaddr {
         self
     }
 
+    /// Ensures the `Multiaddr` is a `/p2p/...` address for the given peer.
+    ///
+    /// If the given address is already a `p2p` address for the given peer,
+    /// i.e. the last encapsulated protocol is `/p2p/<peer-id>`, this is a no-op.
+    ///
+    /// If the given address is already a `p2p` address for a different peer
+    /// than the one given, the peer-id is replaced with the one given.
```

Comment:
I don't think we should do this silently as this could cause very weird behaviour down the line.

I'd propose the following API instead:

```rust
/// Pushes a `/p2p` [`Protocol`] to this address unless it already has one.
///
/// This is a no-op if the address already ends with `/p2p` **even if the embedded [`PeerId`] refers to a different peer.**
/// Thus, after calling this function, the address is guaranteed to end with `/p2p` but it might not refer to the provided [`PeerId`].
pub fn push_p2p(&mut self, peer_id: PeerId) {
	todo!()
}
```

We could also make it a `with_p2p` and take `self` by value instead. That might be more ergonomic to use.

Most importantly though, I think we should change the behaviour and not touch any existing `/p2p` suffixes.

Reply:
> We could also make it a `with_p2p` and take `self` by value instead. That might be more ergonomic to use.

Slightly favoring this immutable API.

> Most importantly though, I think we should change the behaviour and not touch any existing `/p2p` suffixes.

That would require every user to check the `PeerId` in `/p2p` after the call, right? E.g. `libp2p-identity` when receiving a multiaddr from a remote, should check whether the `PeerId` in the reported multiaddr, if any, matches the one of the remote node.

How about keeping the signature and behavior as is, but logging on debug in case an existing `/p2p` is replaced?

Reply:
> How about keeping the signature and behavior as is, but logging on debug in case an existing `/p2p` is replaced?

I just wonder what the value of that functionality is?

If we don't "trust" that the remote is providing us with the correct address, then we should build a different parsing function instead.

This function should IMO more have semantics that are idempotent but safe. Replacing a `/p2p` is not safe but changes the meaning of the address entirely.

Reply:
> then we should build a different parsing function instead.

Are you thinking of something along the lines of the below. Adding the peer ID if none exists. Validating an existing peer ID if one does exist @thomaseizinger?

``` rust
fn parse_with_peer_id(m: &[u8], expected_peer_id: PeerId) -> Result<Multiaddr, _>
```

Reply:
Yeah but I see this as an orthogonal problem. In my eyes, `with_p2p` is primarily for cases where a node deals with its own addresses, not parsing someone else's.

If for whatever reason an existing but different `/p2p` slips in there, replacing it would likely lead to weird bugs down the line.

I think it makes sense to have both but the new parse function would likely only be used in identify, right?

Reply:
> Most importantly though, I think we should change the behaviour and not touch any existing `/p2p` suffixes.

Like @mxinden said, doing so would make the function `push_p2p` not really useful since it would be no different than `push(Protocol::P2p(peer_id))`. The objective of the function we are trying to specify is to provide an uniformed `multiaddr` whether there were `/p2p` suffix or not.

> This function should IMO more have semantics that are idempotent but safe. Replacing a `/p2p` is not safe but changes the meaning of the address entirely.

I think I understand what you mean, and so your reluctance to replacing the `peer_id` in a `multiaddr`. To be sure, is it because `PeerId` are not fixed over time and could change quickly ? But if that so, what is the point to have `p2p` in the `multiaddr` if we cannot trust that a `PeerId` will provide us with a `multiaddr` referencing its own `PeerId` or that it will actually be reachable at the `multiaddr` we stored in our cache ?

There is also the case that end users of the `libp2p` could use the function [`add_address`](https://github.com/libp2p/rust-libp2p/blob/e54580525dea8ee6e0ace1feaa5e46778bc5aaba/protocols/kad/src/behaviour.rs#L527) of the `Kademlia behaviour` to append addresses that could be wrong (`peer_id` in the `multiaddr` not matching the actual `peer_id` we are registering this address to). I agree that it might be the responsibility of the end user to not mess up the routing table but it might still be a good idea to be a little bit resilient against such things.

Would stripping `p2p` suffix instead of happening it simplify this problem ? Since we almost always have the corresponding `peer_id` when working with a `multiaddr`. On the other hand, I fear that doing so may cause security issue since we will not ensure that the `PeerId` which was contained in the `multiaddr` is the actual peer we are talking to.

I am genuinely asking. I admit I do not have the answer :joy:

Reply:
> > Most importantly though, I think we should change the behaviour and not touch any existing `/p2p` suffixes.
> 
> Like @mxinden said, doing so would make the function `push_p2p` not really useful since it would be no different than `push(Protocol::P2p(peer_id))`. The objective of the function we are trying to specify is to provide an uniformed `multiaddr` whether there were `/p2p` suffix or not.

Why is it not useful? `.push(Protocol::P2p(peer_id))` will give you **two** `/p2p` protocols in an address if it already has one. The key different here is that it would be idempotent.

> > This function should IMO more have semantics that are idempotent but safe. Replacing a `/p2p` is not safe but changes the meaning of the address entirely.
> 
> I think I understand what you mean, and so your reluctance to replacing the `peer_id` in a `multiaddr`. To be sure, is it because `PeerId` are not fixed over time and could change quickly ?

No. The reason I don't want to replace `/p2p` protocols is because I don't see a usecase where it makes sense.

Let's say you get a multiaddress from someone: `/ip4/10.0.0.1/tcp/8080/p2p/xyz1`. Now you call `to_p2p_terminated` on it with the `abc1` as the `PeerId`. The resulting multiaddr is now garbage because the peer at `/ip4/10.0.0.1/tcp/8080` does not have identify as `abc1` but as `xyz1`. What is the point of this functionality?

On the other hand, if we have `with_p2p`, we can **always** safely call it on a multiaddr and trust that it will not turn the multiaddress into gargabe. If the address has a `/p2p` suffix, we don't do anything. If it doesn't have one, we add one onto it.

---

Instead of a new API, we could also consider modifying the behaviour of `push`. I think it is quite easy to figure out, which protocols can be duplicated and which ones not. In fact, the only protocol I know of that may be repeated numerous times is `/certhash`. Any other protocol being duplicated results in a useless address. Or am I missing something?

Making the existing `push` function smarter would allow us to make this more convenient without adding another API.

Reply:
Ok. My bad, I had misunderstood what you proposed. Thanks for the clarification.

> Making the existing push function smarter would allow us to make this more convenient without adding another API.

I think this is a great idea. However, I would still prefer something that alerts the caller that the result is not what he'd expect instead of just doing nothing. Would you be open to the following implementation ?

```rust
// on Multiaddr
fn try_with(self, protocol: Protocol) -> Result<Multiaddr, Multiaddr> {
    match multiaddr.iter().last() {
        Some(same protocol and same content) -> Ok(self),
        Some(same protocol but different content) -> Err(self),
        None => Ok(self.with(protocol))
    }
}
```

The only question is about protocols that could be called numerous times like you said : should this function handle them differently or should we let the caller decide if he wants to force push (by using `push / with`) ?  
```rust
// an example of what could be done if the `try_with` function 
// handled "can be duplicated" protocols differently


// on Protocol 
fn can_be_duplicated(&self) -> bool {}

// on Multiaddr
fn try_with(self, protocol: Protocol) -> Result<Multiaddr, Multiaddr> {
    if protocol.can_be_duplicated() {
        return Ok(self.with(protocol);
    }

    match multiaddr.iter().last() {
        Some(same protocol and same content) -> Ok(self),
        Some(same protocol but different content) -> Err(self),
        None => Ok(self.with(protocol))
    }
}
```

I personally would be in favor of letting the caller decide if he wants to force push.

We also could returned more advance error : a tuple containing the original `multiaddr` and the `multiaddr` with the last protocol being replaced by the provided one : `Result<Multiaddr, (Multiaddr, Multiaddr)>`. What would you think about that ?


_To be coherent with what has already been done, I will of course implement a `try_push` and `try_with` function_

Reply:
> ```rust
> // an example of what could be done if the `try_with` function 
> // handled "can be duplicated" protocols differently
> 
> 
> // on Protocol 
> fn can_be_duplicated(&self) -> bool {}
> 
> // on Multiaddr
> fn try_with(self, protocol: Protocol) -> Result<Multiaddr, Multiaddr> {
>     if protocol.can_be_duplicated() {
>         return Ok(self.with(protocol);
>     }
> 
>     match multiaddr.iter().last() {
>         Some(same protocol and same content) -> Ok(self),
>         Some(same protocol but different content) -> Err(self),
>         None => Ok(self.with(protocol))
>     }
> }
> ```

This is great but can be simplified further! I cannot think of a situation in which I want to append _the exact same_ protocol again, e.g. `/tcp/1000/tcp/1000` does not make any sense. Neither does `/certhash/abc/certhash/abc`. It is kind of obvious when you think about it: It would be duplicated information so it is always redundant.

Thus, the actual check we would have to make is:

- if protocol `==` with new one, do nothing
- if different _kind_ of protocol, push
- if same kind but different protocol, do special handling based on protocol, e.g. afaik, only `/certhash` makes sense to be repeated

> > Making the existing push function smarter would allow us to make this more convenient without adding another API.
> 
> I think this is a great idea. However, I would still prefer something that alerts the caller that the result is not what he'd expect instead of just doing nothing

Yeah, I considered an API that "fails" when an existing `PeerId` is discovered that is different from the one we want to push. But when thinking about how I would use that API, I couldn't really come up with a meaningful error handling strategy that I'd apply.

What would you do? Log the error? What would be your log message? What should an operator of your software do in response to seeing the log message?

In general, I agree with you that it is good design to surface such errors to the caller. However, I think this is one of these cases where we end up with a lot of "can never happen" errors.

Hence I think it makes to not make this an error but make `with` opinionated based on which protocol is being pushed. This will do the right thing in 95% of all usecases. If you actually want to replace and existing `/p2p` with a different one, then you can always use a combination of `.iter()`, `.pop()` and `.with()`.

Something like this (untested):

```rust
    /// Like [`Multiaddr::push`] but idempotent depending on the particular [`Protocol`] pushed.
    pub fn with_idempotent(mut self, p: Protocol<'_>) -> Self {
        match (self.iter().last(), &p) {
            (Some(existing), new) if &existing == new => {
                return self // Never append the exact same protocol
            }
            (Some(Protocol::Certhash(_)), Protocol::Certhash(_)) => {
                // Repeating /certhash is allowed as long as they are different.
            }
            (Some(existing), new) if existing.tag() == new.tag() => {
                return self // Repeating any other protocol is not allowed.
            }
            (Some(_) | None, _) => {
                // Pushing onto an empty multiaddr or with different tags is allowed.
            }
        }

        self.with(p)
    }
```

Reply:
Ok great. I agree with almost everything you propose. Still do you really not want a `Result` returned ?

I agree that ~95% of the time the error could not occur. But only the caller will know this for sure. So he could do an `.except("cannot fail")` and if there is the slightest possibility that it could fail, he would be able to handle the case.

I agree that a log would not be of any use. That is why I think what is currently being done in the `libp2p` repo is the way to go (returning the original `multiaddr` as an error to give possibility to the caller to handle this `multiaddr` the best way he see fit: dropping it, fixing it, banning it, etc)

https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/lib.rs#L1950-L1965

I will trust your experience with the usage of `multiaddr`, so this should be the last message before I implement the way you think would best fit usage of `multiaddr` :wink:

Reply:
Let's add two functions:

```rust
impl Multiaddr {
	/// Parses a [`Multiaddr`] from the given string, ensuring that it ends with [`Protocol::P2p`] of the provided peer.
	///
	/// If the address does not end with `/p2p`, it is simply appended.
	/// If the address ends with a _different_ peer ID, parsing will fail.
	pub fn from_str_with_p2p(addr: &str, peer: PeerId) -> Result<Self, Error>;
	
	/// Appends the given [`PeerId`] if not yet present at the end of this multiaddress.
	///
	/// Fails if this address ends in a _different_ [`PeerId`].
	/// In that case, the original, unmodified address is returned.
	pub fn with_p2p(mut self, peer: PeerId) -> Result<Self, Self>;
}
```

We can always generalize them later to something like `with_idempotent` above. For now, I think these should cover us well.

Reply:
It's done. Thanks to the function signatures and documentation you provided it was quite easy in the end 😅

Hope it does match your expectation. Thank you for your patience and insights.